README.md
# prosemirror-collab

[ [**WEBSITE**](https://prosemirror.net) | [**ISSUES**](https://github.com/prosemirror/prosemirror/issues) | [**FORUM**](https://discuss.prosemirror.net) | [**GITTER**](https://gitter.im/ProseMirror/prosemirror) | [**CHANGELOG**](https://github.com/ProseMirror/prosemirror-collab/blob/master/CHANGELOG.md) ]

This is a [core module](https://prosemirror.net/docs/ref/#collab) of [ProseMirror](https://prosemirror.net).
ProseMirror is a well-behaved rich semantic content editor based on
contentEditable, with support for collaborative editing and custom
document schemas.

This [module](https://prosemirror.net/docs/ref/#collab) implements a
plugin that helps track and merge changes for
[collaborative editing](https://prosemirror.net/docs/guide/#collab).

The [project page](https://prosemirror.net) has more information, a
number of [examples](https://prosemirror.net/examples/) and the
[documentation](https://prosemirror.net/docs/).

This code is released under an
[MIT license](https://github.com/prosemirror/prosemirror/tree/master/LICENSE).
There's a [forum](http://discuss.prosemirror.net) for general
discussion and support requests, and the
[Github bug tracker](https://github.com/prosemirror/prosemirror/issues)
is the place to report issues.

We aim to be an inclusive, welcoming community. To make that explicit,
we have a [code of
conduct](http://contributor-covenant.org/version/1/1/0/) that applies
to communication around the project.


CHANGELOG.md
## 1.3.0 (2022-05-30)

### New features

Include TypeScript type declarations.

## 1.2.2 (2019-11-20)

### Bug fixes

Rename ES module files to use a .js extension, since Webpack gets confused by .mjs

## 1.2.1 (2019-11-19)

### Bug fixes

The file referred to in the package's `module` field now is compiled down to ES5.

## 1.2.0 (2019-11-08)

### New features

Add a `module` field to package json file.

## 1.1.2 (2019-05-29)

### Bug fixes

Fix an issue where in `mapSelectionBackward` mode, the plugin flipped the head and anchor of the selection, leading to selection glitches during collaborative editing.

## 1.1.1 (2018-10-09)

### Bug fixes

Fix issue where `mapSelectionBackward` didn't work because of a typo.

## 1.1.0 (2018-08-21)

### New features

[`receiveTransaction`](https://prosemirror.net/docs/ref/#collab.receiveTransaction) now supports a `mapSelectionBackward` option that makes it so that text selections are mapped to stay in place when remote changes insert content at their position.

## 0.19.0 (2017-03-16)

### New features

You can now use strings (as well as numbers) as client IDs (this already worked, but now the documentation reflects this).

## 0.18.0 (2017-02-24)

### New features

[`sendableSteps`](https://prosemirror.net/docs/ref/version/0.18.0.html#collab.sendableSteps) now also returns information about the original transactions that produced the steps.

## 0.11.0 (2016-09-21)

### Breaking changes

Moved into a separate module.

Interface [adjusted](https://prosemirror.net/docs/ref/version/0.11.0.html#collab) to work with the new
[plugin](https://prosemirror.net/docs/ref/version/0.11.0.html#state.Plugin) system.

### New features

When receiving changes, the module now
[generates](https://prosemirror.net/docs/ref/version/0.11.0.html#collab.receiveAction) a regular
[transform action](https://prosemirror.net/docs/ref/version/0.11.0.html#state.TransformAction) instead of hard-setting
the editor's document. This solves problematic corner cases for code
keeping track of the document by listening to transform actions.



test/test-collab.js
const {EditorState, Selection} = require("prosemirror-state")
const {history, undo, redo, closeHistory} = require("prosemirror-history")
const {schema, eq, doc, p} = require("prosemirror-test-builder")
const ist = require("ist")

const {collab, receiveTransaction, sendableSteps} = require("..")

const histPlugin = history()

class DummyServer {
  constructor(doc, n = 2) {
    this.states = []
    this.plugins = []
    for (let i = 0; i < n; i++) {
      let plugin = collab()
      this.plugins.push(plugin)
      this.states.push(EditorState.create({doc, schema, plugins: [histPlugin, plugin]}))
    }
    this.steps = []
    this.clientIDs = []
    this.delayed = []
  }

  sync(n) {
    let state = this.states[n], version = this.plugins[n].getState(state).version
    if (version != this.steps.length)
      this.states[n] = state.apply(receiveTransaction(state, this.steps.slice(version), this.clientIDs.slice(version)))
  }

  send(n) {
    let sendable = sendableSteps(this.states[n])
    if (sendable && sendable.version == this.steps.length) {
      this.steps = this.steps.concat(sendable.steps)
      for (let i = 0; i < sendable.steps.length; i++) this.clientIDs.push(sendable.clientID)
    }
  }

  broadcast(n) {
    if (this.delayed.indexOf(n) > -1) return
    this.sync(n)
    this.send(n)
    for (let i = 0; i < this.states.length; i++) if (i != n) this.sync(i)
  }

  update(n, f) {
    this.states[n] = this.states[n].apply(f(this.states[n]))
    this.broadcast(n)
  }

  type(n, text, pos) {
    this.update(n, s => s.tr.insertText(text, pos || s.selection.head))
  }

  undo(n) {
    undo(this.states[n], tr => this.update(n, () => tr))
  }

  redo(n) {
    redo(this.states[n], tr => this.update(n, () => tr))
  }

  conv(d) {
    if (typeof d == "string") d = doc(p(d))
    this.states.forEach(state => ist(state.doc, d, eq))
  }

  delay(n, f) {
    this.delayed.push(n)
    f()
    this.delayed.pop()
    this.broadcast(n)
  }
}

function sel(near) {
  return s => s.tr.setSelection(Selection.near(s.doc.resolve(near)))
}

describe("collab", () => {
  it("converges for simple changes", () => {
    let s = new DummyServer
    s.type(0, "hi")
    s.type(1, "ok", 3)
    s.type(0, "!", 5)
    s.type(1, "...", 1)
    s.conv("...hiok!")
  })

  it("converges for multiple local changes", () => {
    let s = new DummyServer
    s.type(0, "hi")
    s.delay(0, () => {
      s.type(0, "A")
      s.type(1, "X")
      s.type(0, "B")
      s.type(1, "Y")
    })
    s.conv("hiXYAB")
  })

  it("converges with three peers", () => {
    let s = new DummyServer(null, 3)
    s.type(0, "A")
    s.type(1, "U")
    s.type(2, "X")
    s.type(0, "B")
    s.type(1, "V")
    s.type(2, "C")
    s.conv("AUXBVC")
  })

  it("converges with three peers with multiple steps", () => {
    let s = new DummyServer(null, 3)
    s.type(0, "A")
    s.delay(1, () => {
      s.type(1, "U")
      s.type(2, "X")
      s.type(0, "B")
      s.type(1, "V")
      s.type(2, "C")
    })
    s.conv("AXBCUV")
  })

  it("supports undo", () => {
    let s = new DummyServer
    s.type(0, "A")
    s.type(1, "B")
    s.type(0, "C")
    s.undo(1)
    s.conv("AC")
    s.type(1, "D")
    s.type(0, "E")
    s.conv("ACDE")
  })

  it("supports redo", () => {
    let s = new DummyServer
    s.type(0, "A")
    s.type(1, "B")
    s.type(0, "C")
    s.undo(1)
    s.redo(1)
    s.type(1, "D")
    s.type(0, "E")
    s.conv("ABCDE")
  })

  it("supports deep undo", () => {
    let s = new DummyServer(doc(p("hello"), p("bye")))
    s.update(0, sel(6))
    s.update(1, sel(11))
    s.type(0, "!")
    s.type(1, "!")
    s.update(0, s => closeHistory(s.tr))
    s.delay(0, () => {
      s.type(0, " ...")
      s.type(1, " ,,,")
    })
    s.update(0, s => closeHistory(s.tr))
    s.type(0, "*")
    s.type(1, "*")
    s.undo(0)
    s.conv(doc(p("hello! ..."), p("bye! ,,,*")))
    s.undo(0)
    s.undo(0)
    s.conv(doc(p("hello"), p("bye! ,,,*")))
    s.redo(0)
    s.redo(0)
    s.redo(0)
    s.conv(doc(p("hello! ...*"), p("bye! ,,,*")))
    s.undo(0)
    s.undo(0)
    s.conv(doc(p("hello!"), p("bye! ,,,*")))
    s.undo(1)
    s.conv(doc(p("hello!"), p("bye")))
  })

  it("support undo with clashing events", () => {
    let s = new DummyServer(doc(p("hello")))
    s.update(0, sel(6))
    s.type(0, "A")
    s.delay(0, () => {
      s.type(0, "B", 4)
      s.type(0, "C", 5)
      s.type(0, "D", 1)
      s.update(1, s => s.tr.delete(2, 5))
    })
    s.conv("DhoA")
    s.undo(0)
    s.undo(0)
    s.conv("ho")
    ist(s.states[0].selection.head, 3)
  })

  it("handles conflicting steps", () => {
    let s = new DummyServer(doc(p("abcde")))
    s.delay(0, () => {
      s.update(0, s => s.tr.delete(3, 4))
      s.type(0, "x")
      s.update(1, s => s.tr.delete(2, 5))
    })
    s.undo(0)
    s.undo(0)
    s.conv(doc(p("ae")))
  })

  it("can undo simultaneous typing", () => {
    let s = new DummyServer(doc(p("A"), p("B")))
    s.update(0, sel(2))
    s.update(1, sel(5))
    s.delay(0, () => {
      s.type(0, "1")
      s.type(0, "2")
      s.type(1, "x")
      s.type(1, "y")
    })
    s.conv(doc(p("A12"), p("Bxy")))
    s.undo(0)
    s.conv(doc(p("A"), p("Bxy")))
    s.undo(1)
    s.conv(doc(p("A"), p("B")))
  })
})


test/test-collab.ts
import {EditorState, Selection, Plugin, Transaction} from "prosemirror-state"
import {history, undo, redo, closeHistory} from "prosemirror-history"
import {Node} from "prosemirror-model"
import {Step} from "prosemirror-transform"
import {schema, eq, doc, p} from "prosemirror-test-builder"
import ist from "ist"

import {collab, receiveTransaction, sendableSteps} from "prosemirror-collab"

const histPlugin = history()

class DummyServer {
  states: EditorState[] = []
  plugins: Plugin[] = []
  steps: Step[] = []
  clientIDs: number[] = []
  delayed: number[] = []

  constructor(doc?: Node, n = 2) {
    for (let i = 0; i < n; i++) {
      let plugin = collab()
      this.plugins.push(plugin)
      this.states.push(EditorState.create({doc, schema, plugins: [histPlugin, plugin]}))
    }
  }

  sync(n: number) {
    let state = this.states[n], version = this.plugins[n].getState(state).version
    if (version != this.steps.length)
      this.states[n] = state.apply(receiveTransaction(state, this.steps.slice(version), this.clientIDs.slice(version)))
  }

  send(n: number) {
    let sendable = sendableSteps(this.states[n])
    if (sendable && sendable.version == this.steps.length) {
      this.steps = this.steps.concat(sendable.steps)
      for (let i = 0; i < sendable.steps.length; i++) this.clientIDs.push(sendable.clientID as number)
    }
  }

  broadcast(n: number) {
    if (this.delayed.indexOf(n) > -1) return
    this.sync(n)
    this.send(n)
    for (let i = 0; i < this.states.length; i++) if (i != n) this.sync(i)
  }

  update(n: number, f: (state: EditorState) => Transaction) {
    this.states[n] = this.states[n].apply(f(this.states[n]))
    this.broadcast(n)
  }

  type(n: number, text: string, pos?: number) {
    this.update(n, s => s.tr.insertText(text, pos == null ? s.selection.head : pos))
  }

  undo(n: number) {
    undo(this.states[n], tr => this.update(n, () => tr))
  }

  redo(n: number) {
    redo(this.states[n], tr => this.update(n, () => tr))
  }

  conv(d: Node | string) {
    if (typeof d == "string") d = doc(p(d))
    this.states.forEach(state => ist(state.doc, d, eq))
  }

  delay(n: number, f: () => void) {
    this.delayed.push(n)
    f()
    this.delayed.pop()
    this.broadcast(n)
  }
}

function sel(near: number) {
  return (s: EditorState) => s.tr.setSelection(Selection.near(s.doc.resolve(near)))
}

describe("collab", () => {
  it("converges for simple changes", () => {
    let s = new DummyServer
    s.type(0, "hi")
    s.type(1, "ok", 3)
    s.type(0, "!", 5)
    s.type(1, "...", 1)
    s.conv("...hiok!")
  })

  it("converges for multiple local changes", () => {
    let s = new DummyServer
    s.type(0, "hi")
    s.delay(0, () => {
      s.type(0, "A")
      s.type(1, "X")
      s.type(0, "B")
      s.type(1, "Y")
    })
    s.conv("hiXYAB")
  })

  it("converges with three peers", () => {
    let s = new DummyServer(undefined, 3)
    s.type(0, "A")
    s.type(1, "U")
    s.type(2, "X")
    s.type(0, "B")
    s.type(1, "V")
    s.type(2, "C")
    s.conv("AUXBVC")
  })

  it("converges with three peers with multiple steps", () => {
    let s = new DummyServer(undefined, 3)
    s.type(0, "A")
    s.delay(1, () => {
      s.type(1, "U")
      s.type(2, "X")
      s.type(0, "B")
      s.type(1, "V")
      s.type(2, "C")
    })
    s.conv("AXBCUV")
  })

  it("supports undo", () => {
    let s = new DummyServer
    s.type(0, "A")
    s.type(1, "B")
    s.type(0, "C")
    s.undo(1)
    s.conv("AC")
    s.type(1, "D")
    s.type(0, "E")
    s.conv("ACDE")
  })

  it("supports redo", () => {
    let s = new DummyServer
    s.type(0, "A")
    s.type(1, "B")
    s.type(0, "C")
    s.undo(1)
    s.redo(1)
    s.type(1, "D")
    s.type(0, "E")
    s.conv("ABCDE")
  })

  it("supports deep undo", () => {
    let s = new DummyServer(doc(p("hello"), p("bye")))
    s.update(0, sel(6))
    s.update(1, sel(11))
    s.type(0, "!")
    s.type(1, "!")
    s.update(0, s => closeHistory(s.tr))
    s.delay(0, () => {
      s.type(0, " ...")
      s.type(1, " ,,,")
    })
    s.update(0, s => closeHistory(s.tr))
    s.type(0, "*")
    s.type(1, "*")
    s.undo(0)
    s.conv(doc(p("hello! ..."), p("bye! ,,,*")))
    s.undo(0)
    s.undo(0)
    s.conv(doc(p("hello"), p("bye! ,,,*")))
    s.redo(0)
    s.redo(0)
    s.redo(0)
    s.conv(doc(p("hello! ...*"), p("bye! ,,,*")))
    s.undo(0)
    s.undo(0)
    s.conv(doc(p("hello!"), p("bye! ,,,*")))
    s.undo(1)
    s.conv(doc(p("hello!"), p("bye")))
  })

  it("support undo with clashing events", () => {
    let s = new DummyServer(doc(p("hello")))
    s.update(0, sel(6))
    s.type(0, "A")
    s.delay(0, () => {
      s.type(0, "B", 4)
      s.type(0, "C", 5)
      s.type(0, "D", 1)
      s.update(1, s => s.tr.delete(2, 5))
    })
    s.conv("DhoA")
    s.undo(0)
    s.undo(0)
    s.conv("ho")
    ist(s.states[0].selection.head, 3)
  })

  it("handles conflicting steps", () => {
    let s = new DummyServer(doc(p("abcde")))
    s.delay(0, () => {
      s.update(0, s => s.tr.delete(3, 4))
      s.type(0, "x")
      s.update(1, s => s.tr.delete(2, 5))
    })
    s.undo(0)
    s.undo(0)
    s.conv(doc(p("ae")))
  })

  it("can undo simultaneous typing", () => {
    let s = new DummyServer(doc(p("A"), p("B")))
    s.update(0, sel(2))
    s.update(1, sel(5))
    s.delay(0, () => {
      s.type(0, "1")
      s.type(0, "2")
      s.type(1, "x")
      s.type(1, "y")
    })
    s.conv(doc(p("A12"), p("Bxy")))
    s.undo(0)
    s.conv(doc(p("A"), p("Bxy")))
    s.undo(1)
    s.conv(doc(p("A"), p("B")))
  })
})


test/test-rebase.js
const {Transform} = require("prosemirror-transform")
const ist = require("ist")
const {schema, eq, doc, blockquote, p, li, ul, em} = require("prosemirror-test-builder")

const {rebaseSteps} = require("..")

function runRebase(transforms, expected) {
  let start = transforms[0].before, full = new Transform(start)
  transforms.forEach(transform => {
    let rebased = new Transform(transform.doc)
    let start = transform.steps.length + full.steps.length
    rebaseSteps(transform.steps.map((s, i) => ({step: s, inverted: s.invert(transform.docs[i])})),
                full.steps, rebased)
    for (let i = start; i < rebased.steps.length; i++) full.step(rebased.steps[i])
  })

  ist(full.doc, expected, eq)

  for (let tag in start.tag) {
    let mapped = full.mapping.mapResult(start.tag[tag])

    let exp = expected.tag[tag]
    if (mapped.deleted) {
      if (exp) throw new Error("Tag " + tag + " was unexpectedly deleted")
    } else {
      if (!exp) throw new Error("Tag " + tag + " is not actually deleted")
      ist(mapped.pos, exp)
    }
  }
}

function permute(array) {
  if (array.length < 2) return [array]
  let result = []
  for (let i = 0; i < array.length; i++) {
    let others = permute(array.slice(0, i).concat(array.slice(i + 1)))
    for (let j = 0; j < others.length; j++)
      result.push([array[i]].concat(others[j]))
  }
  return result
}

function rebase(doc, ...clients) {
  let expected = clients.pop()
  runRebase(clients.map(cl => cl(new Transform(doc))), expected)
}

function rebase$(doc, ...clients) {
  let expected = clients.pop()
  permute(clients.map(cl => cl(new Transform(doc)))).forEach(transforms => runRebase(transforms, expected))
}

function type(tr, pos, text) {
  return tr.replaceWith(pos, pos, schema.text(text))
}

function wrap(tr, pos, type) {
  let $pos = tr.doc.resolve(pos)
  return tr.wrap($pos.blockRange($pos), [{type: schema.nodes[type]}])
}

describe("rebaseSteps", () => {
  it("supports concurrent typing", () => {
    rebase$(doc(p("h<1>ell<2>o")),
            tr => type(tr, 2, "X"),
            tr => type(tr, 5, "Y"),
            doc(p("hX<1>ellY<2>o")))
  })

  it("support multiple concurrently typed chars", () => {
    rebase$(doc(p("h<1>ell<2>o")),
            tr => type(type(type(tr, 2, "X"), 3, "Y"), 4, "Z"),
            tr => type(type(tr, 5, "U"), 6, "V"),
            doc(p("hXYZ<1>ellUV<2>o")))
  })

  it("supports three concurrent typers", () => {
    rebase$(doc(p("h<1>ell<2>o th<3>ere")),
            tr => type(tr, 2, "X"),
            tr => type(tr, 5, "Y"),
            tr => type(tr, 9, "Z"),
            doc(p("hX<1>ellY<2>o thZ<3>ere")))
  })

  it("handles wrapping of changed blocks", () => {
    rebase$(doc(p("<1>hell<2>o<3>")),
            tr => type(tr, 5, "X"),
            tr => wrap(tr, 1, "blockquote"),
            doc(blockquote(p("<1>hellX<2>o<3>"))))
  })

  it("handles insertions in deleted content", () => {
    rebase$(doc(p("hello<1> wo<2>rld<3>!")),
            tr => tr.delete(6, 12),
            tr => type(tr, 9, "X"),
            doc(p("hello<3>!")))
  })

  it("allows deleting the same content twice", () => {
    rebase(doc(p("hello<1> wo<2>rld<3>!")),
           tr => tr.delete(6, 12),
           tr => tr.delete(6, 12),
           doc(p("hello<3>!")))
  })

  it("isn't confused by joining a block that's being edited", () => {
    rebase$(doc(ul(li(p("one")), "<1>", li(p("tw<2>o")))),
            tr => type(tr, 12, "A"),
            tr => tr.join(8),
            doc(ul(li(p("one"), p("twA<2>o")))))
  })

  it("supports typing concurrently with marking", () => {
    rebase(doc(p("hello <1>wo<2>rld<3>")),
           tr => tr.addMark(7, 12, schema.mark("em")),
           tr => type(tr, 9, "_"),
           doc(p("hello <1>", em("wo"), "_<2>", em("rld<3>"))))
  })

  it("doesn't unmark marks added concurrently", () => {
    rebase(doc(p(em("<1>hello"), " world<2>")),
           tr => tr.addMark(1, 12, schema.mark("em")),
           tr => tr.removeMark(1, 12, schema.mark("em")),
           doc(p("<1>hello", em(" world<2>"))))
  })

  it("doesn't mark concurrently unmarked text", () => {
    rebase(doc(p("<1>hello ", em("world<2>"))),
           tr => tr.removeMark(1, 12, schema.mark("em")),
           tr => tr.addMark(1, 12, schema.mark("em")),
           doc(p(em("<1>hello "), "world<2>")))
  })

  it("deletes inserts in replaced context", () => {
    rebase(doc(p("b<before>efore"), blockquote(ul(li(p("o<1>ne")), li(p("t<2>wo")), li(p("thr<3>ee")))), p("a<after>fter")),
           tr => tr.replace(tr.doc.tag[1], tr.doc.tag[3], doc(p("a"), blockquote(p("b")), p("c")).slice(2, 9)),
           tr => type(tr, tr.doc.tag[2], "ayay"),
           doc(p("b<before>efore"), blockquote(ul(li(p("o"), blockquote(p("b")), p("<3>ee")))), p("a<after>fter")))
  })

  it("maps through inserts", () => {
    rebase$(doc(p("X<1>X<2>X")),
            tr => type(tr, 2, "hello"),
            tr => type(tr, 3, "goodbye").delete(4, 7),
            doc(p("Xhello<1>Xgbye<2>X")))
  })

  it("handle concurrent removal of blocks", () => {
    rebase(doc(p("a"), "<1>", p("b"), "<2>", p("c")),
           tr => tr.delete(tr.doc.tag[1], tr.doc.tag[2]),
           tr => tr.delete(tr.doc.tag[1], tr.doc.tag[2]),
           doc(p("a"), "<2>", p("c")))
  })

  it("discards edits in removed blocks", () => {
    rebase$(doc(p("a"), "<1>", p("b<2>"), "<3>", p("c")),
            tr => tr.delete(tr.doc.tag[1], tr.doc.tag[3]),
            tr => type(tr, tr.doc.tag[2], "ay"),
            doc(p("a"), "<3>", p("c")))
  })

  it("preserves double block inserts", () => {
    rebase(doc(p("a"), "<1>", p("b")),
           tr => tr.replaceWith(3, 3, schema.node("paragraph")),
           tr => tr.replaceWith(3, 3, schema.node("paragraph")),
           doc(p("a"), p(), p(), "<1>", p("b")))
  })
})


index.js
const {Plugin} = require("../state")
const RopeSequence = require("rope-sequence")

const {rebaseSteps} = require("./rebase")
exports.rebaseSteps = rebaseSteps

// This state field accumulates changes that have to be sent to the
// central authority in the collaborating group and makes it possible
// to integrate changes made by peers into our local document. It is
// defined by the plugin, and will be available as the `collab` field
// in the resulting editor state.
class CollabState {
  constructor(version, unconfirmed) {
    // : number
    // The version number of the last update received from the central
    // authority. Starts at 0 or the value of the `version` property
    // in the option object, for the editor's value when the option
    // was enabled.
    this.version = version

    // : RopeSequence<{step: Step, inverted: Step}>
    // The local steps that havent been successfully sent to the
    // server yet.
    this.unconfirmed = unconfirmed
  }
}

function unconfirmedFrom(transform, start = 0) {
  let add = []
  for (let i = start; i < transform.steps.length; i++)
    add.push({step: transform.steps[i],
              inverted: transform.steps[i].invert(transform.docs[i])})
  return add
}

// Pushes a set of steps (received from the central authority) into
// the editor state (which should have the collab plugin enabled).
// Will recognize its own changes, and confirm unconfirmed steps as
// appropriate. Remaining unconfirmed steps will be rebased over
// remote steps.
function makeReceiveAction(state, steps, clientIDs, ourID) {
  let collab = state.collab
  let version = collab.version + steps.length

  // Find out which prefix of the steps originated with us
  let ours = 0
  while (ours < clientIDs.length && clientIDs[ours] == ourID) ++ours
  let unconfirmed = collab.unconfirmed.slice(ours)
  steps = ours ? steps.slice(ours) : steps

  // If all steps originated with us, we're done.
  if (!steps.length)
    return {type: "collabConfirm", collabState: new CollabState(version, unconfirmed)}

  let nUnconfirmed = unconfirmed.length
  let transform = state.tr
  if (nUnconfirmed) {
    rebaseSteps(transform, unconfirmed.map(s => s.step), unconfirmed.map(s => s.inverted), steps)
  } else {
    for (let i = 0; i < steps.length; i++) transform.step(steps[i])
  }

  unconfirmed = RopeSequence.from(unconfirmedFrom(transform, nUnconfirmed + steps.length))
  let newCollabState = new CollabState(version, unconfirmed)
  return transform.action({rebased: nUnconfirmed, addToHistory: false, newCollabState, interaction: false})
}

const plugin = new Plugin({
  stateFields: {
    collab: {
      init: (_, state) => new CollabState(plugin.find(state).config.version, RopeSequence.empty),
      applyAction({collab}, action) {
        if (action.type == "transform")
          return action.newCollabState ||
          new CollabState(collab.version, collab.unconfirmed.append(unconfirmedFrom(action.transform)))
        if (action.type == "collabConfirm")
          return action.collabState
        return collab
      }
    }
  },

  config: {
    version: 0,
    clientID: -1
  }
})

// :: (?Object) → Plugin
//
// Creates a plugin that enables the collaborative editing framework
// for the editor.
//
//   config::- An optional set of options
//
//     version:: ?number
//     The starting version number of the collaborative editing.
//     Defaults to 0.
//
//     clientID:: ?number
//     This client's ID. Defaults to a random 32-bit number.
function collab(config) {
  let clientID = config && config.clientID
  if (clientID == null) clientID =  Math.floor(Math.random() * 0xFFFFFFFF)

  let reconf = {clientID}
  if (config && config.version != null) reconf.version = config.version
  return plugin.configure(reconf)
}
exports.collab = collab

// :: (state: EditorState, steps: [Step], clientIDs: [number]) → Action
// Create an action that represents a set of new steps received from
// the authority. Applying this action moves the state forward along
// with the authority's view of the document.
function receiveAction(state, steps, clientIDs) {
  return makeReceiveAction(state, steps, clientIDs, plugin.find(state).config.clientID)
}
exports.receiveAction = receiveAction

// :: (state: EditorState) → ?{version: number, steps: [Step], clientID: number}
// Provides the data describing the editor's unconfirmed steps. The
// version and array of steps are the things you'd send to the
// central authority. Returns null when there is nothing to send.
function sendableSteps(state) {
  if (state.collab.unconfirmed.length == 0) return null
  return {
    version: state.collab.version,
    steps: state.collab.unconfirmed.map(s => s.step),
    clientID: plugin.find(state).config.clientID
  }
}
exports.sendableSteps = sendableSteps


src/collab.ts
import {Plugin, PluginKey, TextSelection, EditorState, Transaction} from "prosemirror-state"
import {Step, Transform} from "prosemirror-transform"

class Rebaseable {
  constructor(
    readonly step: Step,
    readonly inverted: Step,
    readonly origin: Transform
  ) {}
}

/// Undo a given set of steps, apply a set of other steps, and then
/// redo them @internal
export function rebaseSteps(steps: readonly Rebaseable[], over: readonly Step[], transform: Transform) {
  for (let i = steps.length - 1; i >= 0; i--) transform.step(steps[i].inverted)
  for (let i = 0; i < over.length; i++) transform.step(over[i])
  let result = []
  for (let i = 0, mapFrom = steps.length; i < steps.length; i++) {
    let mapped = steps[i].step.map(transform.mapping.slice(mapFrom))
    mapFrom--
    if (mapped && !transform.maybeStep(mapped).failed) {
      transform.mapping.setMirror(mapFrom, transform.steps.length - 1)
      result.push(new Rebaseable(mapped, mapped.invert(transform.docs[transform.docs.length - 1]), steps[i].origin))
    }
  }
  return result
}

// This state field accumulates changes that have to be sent to the
// central authority in the collaborating group and makes it possible
// to integrate changes made by peers into our local document. It is
// defined by the plugin, and will be available as the `collab` field
// in the resulting editor state.
class CollabState {
  constructor(
    // The version number of the last update received from the central
    // authority. Starts at 0 or the value of the `version` property
    // in the option object, for the editor's value when the option
    // was enabled.
    readonly version: number,
    // The local steps that havent been successfully sent to the
    // server yet.
    readonly unconfirmed: readonly Rebaseable[]
  ) {}
}

function unconfirmedFrom(transform: Transform) {
  let result = []
  for (let i = 0; i < transform.steps.length; i++)
    result.push(new Rebaseable(transform.steps[i],
                               transform.steps[i].invert(transform.docs[i]),
                               transform))
  return result
}

const collabKey = new PluginKey("collab")

type CollabConfig = {
  /// The starting version number of the collaborative editing.
  /// Defaults to 0.
  version?: number

  /// This client's ID, used to distinguish its changes from those of
  /// other clients. Defaults to a random 32-bit number.
  clientID?: number | string
}

/// Creates a plugin that enables the collaborative editing framework
/// for the editor.
export function collab(config: CollabConfig = {}) {
  let conf: Required<CollabConfig> = {
    version: config.version || 0,
    clientID: config.clientID == null ? Math.floor(Math.random() * 0xFFFFFFFF) : config.clientID
  }

  return new Plugin({
    key: collabKey,

    state: {
      init: () => new CollabState(conf.version, []),
      apply(tr, collab) {
        let newState = tr.getMeta(collabKey)
        if (newState)
          return newState
        if (tr.docChanged)
          return new CollabState(collab.version, collab.unconfirmed.concat(unconfirmedFrom(tr)))
        return collab
      }
    },

    // @ts-ignore
    config: conf,

    // This is used to notify the history plugin to not merge steps,
    // so that the history can be rebased.
    historyPreserveItems: true
  })
}

/// Create a transaction that represents a set of new steps received from
/// the authority. Applying this transaction moves the state forward to
/// adjust to the authority's view of the document.
export function receiveTransaction(
  state: EditorState,
  steps: readonly Step[],
  clientIDs: readonly (string | number)[],
  options: {
    /// When enabled (the default is `false`), if the current
    /// selection is a [text selection](#state.TextSelection), its
    /// sides are mapped with a negative bias for this transaction, so
    /// that content inserted at the cursor ends up after the cursor.
    /// Users usually prefer this, but it isn't done by default for
    /// reasons of backwards compatibility.
    mapSelectionBackward?: boolean
  } = {}
) {
  // Pushes a set of steps (received from the central authority) into
  // the editor state (which should have the collab plugin enabled).
  // Will recognize its own changes, and confirm unconfirmed steps as
  // appropriate. Remaining unconfirmed steps will be rebased over
  // remote steps.
  let collabState = collabKey.getState(state)
  let version = collabState.version + steps.length
  let ourID: string | number = (collabKey.get(state)!.spec as any).config.clientID

  // Find out which prefix of the steps originated with us
  let ours = 0
  while (ours < clientIDs.length && clientIDs[ours] == ourID) ++ours
  let unconfirmed = collabState.unconfirmed.slice(ours)
  steps = ours ? steps.slice(ours) : steps

  // If all steps originated with us, we're done.
  if (!steps.length)
    return state.tr.setMeta(collabKey, new CollabState(version, unconfirmed))

  let nUnconfirmed = unconfirmed.length
  let tr = state.tr
  if (nUnconfirmed) {
    unconfirmed = rebaseSteps(unconfirmed, steps, tr)
  } else {
    for (let i = 0; i < steps.length; i++) tr.step(steps[i])
    unconfirmed = []
  }

  let newCollabState = new CollabState(version, unconfirmed)
  if (options && options.mapSelectionBackward && state.selection instanceof TextSelection) {
    tr.setSelection(TextSelection.between(tr.doc.resolve(tr.mapping.map(state.selection.anchor, -1)),
                                          tr.doc.resolve(tr.mapping.map(state.selection.head, -1)), -1))
    ;(tr as any).updated &= ~1
  }
  return tr.setMeta("rebased", nUnconfirmed).setMeta("addToHistory", false).setMeta(collabKey, newCollabState)
}

/// Provides data describing the editor's unconfirmed steps, which need
/// to be sent to the central authority. Returns null when there is
/// nothing to send.
///
/// `origins` holds the _original_ transactions that produced each
/// steps. This can be useful for looking up time stamps and other
/// metadata for the steps, but note that the steps may have been
/// rebased, whereas the origin transactions are still the old,
/// unchanged objects.
export function sendableSteps(state: EditorState): {
  version: number,
  steps: readonly Step[],
  clientID: number | string,
  origins: readonly Transaction[]
} | null {
  let collabState = collabKey.getState(state) as CollabState
  if (collabState.unconfirmed.length == 0) return null
  return {
    version: collabState.version,
    steps: collabState.unconfirmed.map(s => s.step),
    clientID: (collabKey.get(state)!.spec as any).config.clientID,
    get origins() {
      return (this as any)._origins || ((this as any)._origins = collabState.unconfirmed.map(s => s.origin))
    }
  }
}

/// Get the version up to which the collab plugin has synced with the
/// central authority.
export function getVersion(state: EditorState): number {
  return collabKey.getState(state).version
}
